ompi/mca/mtl/ofi/mtl_ofi.h

```diff
@@ -282,20 +293,37 @@ ompi_mtl_ofi_send_start(struct mca_mtl_base_module_t *mtl,
         }
     } else {
         ofi_req->completion_count = 1;
-        MTL_OFI_SET_SEND_BITS(match_bits, comm->c_contextid,
-                              comm->c_my_rank, tag, 0);
     }
 
     if (ompi_mtl_ofi.max_inject_size >= length) {
-        MTL_OFI_RETRY_UNTIL_DONE(fi_tinject(ompi_mtl_ofi.ep,
+        if (ompi_mtl_ofi.fi_cq_data) {
+            MTL_OFI_RETRY_UNTIL_DONE(fi_tinjectdata(ompi_mtl_ofi.ep,
+                                            start,
+                                            length,
+                                            comm->c_my_rank,
+                                            endpoint->peer_fiaddr,
+                                            match_bits));
+        } else {
+            MTL_OFI_RETRY_UNTIL_DONE(fi_tinject(ompi_mtl_ofi.ep,
                                             start,
                                             length,
                                             endpoint->peer_fiaddr,
                                             match_bits));
+        }
+
         if (OPAL_UNLIKELY(0 > ret)) {
+            char *fi_api;
+            if (ompi_mtl_ofi.fi_cq_data) {
+                    asprintf( &fi_api, "fi_tinjectddata");
```

Comment:
Any reason not to use a string constant here?  I know this is an error path, and it doesn't really matter, but it might be slightly simpler to:

```c
char *fi_api = ompi_mtl_ofi.fi_cq_data ? " fi_tinjectddata" : "fi_tinject";
```

Reply:
No limitation and indeed looks nicer. Will update.